pauxy/propagation/hubbard_holstein.py

```diff
@@ -0,0 +1,979 @@
+            if abs(walker.weight) > 0:
+                r = numpy.random.random()
+                if r < 0.5:
+                    xi = 0
+                else:
+                    xi = 1
+                vtup = walker.phi[i,:nup] * delta[xi, 0]
+                vtdown = walker.phi[i,nup:] * delta[xi, 1]
+                walker.phi[i,:nup] = walker.phi[i,:nup] + vtup
+                walker.phi[i,nup:] = walker.phi[i,nup:] + vtdown
+                if (self.charge):
+                    walker.weight *= self.charge_factor[xi]
+        
+        if (self.lang_firsov):
+            Dp = numpy.array([numpy.exp(1j*system.gamma*walker.P[i]) for i in range(system.nbasis)])
+            T = numpy.zeros_like(system.T, dtype=numpy.complex128)
+            T[0] = numpy.diag(Dp).dot(system.T[0]).dot(numpy.diag(Dp.T.conj())) + numpy.eye(system.nbasis) * self.onebody_lf
+            T[1] = numpy.diag(Dp).dot(system.T[1]).dot(numpy.diag(Dp.T.conj())) + numpy.eye(system.nbasis) * self.onebody_lf
+            
+            self.bt2 = numpy.array([scipy.linalg.expm(-0.5*self.dt*T[0]),
+                                    scipy.linalg.expm(-0.5*self.dt*T[1])])
+            
+            self.kinetic(walker.phi, system, self.bt2)
+        else:
+            kinetic_real(walker.phi, system, Veph, H1diag=True)
+
+            kinetic_real(walker.phi, system, self.bt2)
+
+        walker.inverse_overlap(trial)
+        
+        # Update walker weight
+        if (self.lang_firsov):
+            walker.ot = walker.calc_otrial(trial.psi) * self.boson_trial.value(walker.P)
+        else:
+            walker.ot = walker.calc_otrial(trial.psi) * self.boson_trial.value(walker.X)
+
+        walker.greens_function(trial)
+        # Constant terms are included in the walker's weight.
+        # (magn, dtheta) = cmath.polar(cmath.exp(cmf+self.dt*eshift))
+        # walker.weight *= magn
+        # walker.phase *= cmath.exp(1j*dtheta)
+
+
+def calculate_overlap_ratio_multi_ghf(walker, delta, trial, i):
+    """Calculate overlap ratio for single site update with GHF trial.
+
+    Parameters
+    ----------
+    walker : walker object
+        Walker to be updated.
+    delta : :class:`numpy.ndarray`
+        Delta updates for single spin flip.
+    trial : trial wavefunctio object
+        Trial wavefunction.
+    i : int
+        Basis index.
+    """
+    nbasis = trial.psi.shape[1] // 2
+    for (idx, G) in enumerate(walker.Gi):
+        guu = G[i,i]
+        gdd = G[i+nbasis,i+nbasis]
+        gud = G[i,i+nbasis]
+        gdu = G[i+nbasis,i]
+        walker.R[idx,0] = (
+            (1+delta[0,0]*guu)*(1+delta[0,1]*gdd) - delta[0,0]*gud*delta[0,1]*gdu
+        )
+        walker.R[idx,1] = (
+            (1+delta[1,0]*guu)*(1+delta[1,1]*gdd) - delta[1,0]*gud*delta[1,1]*gdu
+        )
+    R = numpy.einsum('i,ij,i->j',trial.coeffs,walker.R,walker.ots)/walker.ot
+    return 0.5 * numpy.array([R[0],R[1]])
+
+def calculate_overlap_ratio_multi_det(walker, delta, trial, i):
```

Comment:
There's a lot of copy pasted functions here which probably don't work in for HH in general no?